app/src/main/java/com/google/jetpackcamera/MainActivity.kt

```diff
@@ -101,13 +107,67 @@ class MainActivity : ComponentActivity() {
                             modifier = Modifier.fillMaxSize(),
                             color = MaterialTheme.colorScheme.background
                         ) {
-                            JcaApp(onPreviewViewModel = { previewViewModel = it })
+                            JcaApp(
+                                onPreviewViewModel = { previewViewModel = it },
+                                contentResolver = contentResolver,
+                                imageCaptureUri = uri,
+                                onImageCapture = { event ->
+                                    val shouldFinishWithResult = isExternalImageCapture()
+                                    when (event) {
+                                        is CameraUseCase.ImageCaptureEvent.ImageSaved -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_OK)
+                                                finish()
+                                            }
+                                        }
+
+                                        is CameraUseCase.ImageCaptureEvent.ImageCaptureError -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_CANCELED)
```

Comment:
I'm not sure that's the best UX. I think it could cause a lot of confusion by potentially making errors happening repeatedly within JCA. With the current UX, the user would be forced to examine the RESULT_CANCEL within there own app. I think with this logic, would could consider incorporating the error message or exception in the result intent.

Reply:
Agreed that it's not the best UX, but it may be better than the alternative.
`ACTION_IMAGE_CAPTURE` does not specify any result codes or any extras returned in the result that would denote an error, so we're stuck with the `RESULT_OK` and `RESULT_CANCELED`. Sure we could put an error message or exception into the result, but since it isn't documented anywhere for `ACTION_IMAGE_CAPTURE`, apps can't reasonably investigate the results from an intent because they don't know what to look for. Remember, JCA is only one of many apps that can handle `ACTION_IMAGE_CAPTURED`, so apps won't know specifics of what is returned unless it is part of the documentation.

Since *we* know what the error message is and what it means, we have enough context to show the error in the UI in JCA. It's also possible it could be a transitive error, in which case trying again would be faster if we stayed in the app. If the user navigates away, the `RESULT_CANCELED` is actually the correct error code for the calling app, since it was the user's decision to cancel the request.

Note this is how the AOSP camera app handles this scenario [here](https://cs.android.com/android/platform/superproject/main/+/main:packages/apps/Camera2/src/com/android/camera/PhotoModule.java;l=1289;drc=96ae7ed133f130b84cfa6e71ab6b5c23f806f046).